crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -520,6 +523,7 @@ impl SimpleFaucet {
                     .sign_and_execute_txn(uuid, recipient, coin_id, tx_data, false)
                     .await?;
                 self.metrics.total_coin_requests_succeeded.inc();
+                self.metrics.balance.add(-(total_amount as i64));
```

Comment:
This doesn't match the actual SUI spent by the transaction (e.g. the amount sent, but also the amount spent on the transaction for computation, storage etc), so it will diverge over time.

One way to handle this is to request balance changes when executing the transaction. The balance change for the faucet's address, should match the balance change on the coin, and can be applied to the metric as a delta. I would suggest moving that logic into `sign_and_execute_txn` so you only need to handle it in one place.

Reply:
Funnily enough, requesting balance change in the `sign_and_execute_txn` returns `None`.

```
        match timeout(
            Duration::from_secs(300),
            self.execute_pay_sui_txn_with_retries(&tx, coin_id, recipient, uuid),
        )
        .await
        {
        [ommitted ...]
          Ok(result) {
                if for_batch {
                    self.recycle_gas_coin_for_batch(coin_id, uuid).await;
                } else {
                    self.recycle_gas_coin(coin_id, uuid).await;
                }

                println!("Transaction balance changes: {:?}", result.balance_changes);
```

Results in: 
```
➜  sui git:(faucet_add_total_balance_metric) ✗ sui start --with-faucet --force-regenesis
2024-10-05T01:46:11.897676Z ERROR mysten_metrics::thread_stall_monitor: Thread stalled for 530ms
2024-10-05T01:46:17.861654Z ERROR mysten_metrics::thread_stall_monitor: Thread stall cleared after 6434ms
Transaction balance changes: None
```

Reply:
I was hoping this would work but it does not.

```
                if let Some(ref balances) = result.balance_changes {
                    let sui_used = balances
                        .iter()
                        .find(|balance| balance.owner == self.active_address)
                        .map(|b| b.amount)
                        .unwrap_or(0);
                    self.metrics.balance.add(sui_used as i64);
                }
```

Reply:
Hmm, this is a WaitForLocalExecution thing, I think. (Namely that we don't have it anymore).

- Does it work to fetch the transaction by its digest after you execute the transaction? (Does the codepath that the faucet uses to execute transactions include the logic that tries to simulate WaitForLocalExecution that we added to the SDK, where we poll for the transaction?)
- Alternatively, could you poll for the coin object at the appropriate version and extract the balance from there?

Reply:
Right, forgot about that. Looking at the code, it does use the updated API in the SDK where polling for the tx happens, so then something is failing there.

Reply:
Ah note that this is run on local network, but I'd expect to see the same behaviour regardless of network.

Reply:
After taking a moment to think, I realized that the JSON RPC request options did not include `with_balance_changes`, so we did not get back the needed data. 
Should be good to go now!